### ISSUE TITLE
Potential crossing streams issues in v4 (IKEv2)

### ISSUE THREAD

**Issue by The-Mule** [libreswan/libreswan#1903]
### What happened?

Consider the following scenario:

1.  three nodes, host-to-host tunnels configured between all pairs, ikev2, rsasig peer authentication, auto=route
2. all nodes do --start --asynchronous to all other nodes more or less at the same time, it is a sort of mesh where all tunnels are established in the beginning

With libreswan 4.6 - 4.15 in one of ~20-30 iterations of this scenario at least one IPsec SA won't be established correctly. Once I scale to 10 nodes it is one in ~2 and with 20 or more nodes it always happens. 

Interestingly, downgrading to version 4.5 will make this problem disappear (it is still reproducible but only with more than 30 nodes and a lot of iterations). So some changes introduced later increased this crossing streams probability a lot. Also, the issue seems to go away with version 5. But there are still a lot folks running version 4 for good reasons. 

In the following example, the problem is between node-1 (establishing both tun-3-in-1 and tun-3-out-1) and node-3 (establishing both tun-1-in-1 and tun-1-out-1). Notice that connections tun-3-in-1 and tun-3-out-1 on node-1 matche tun-1-out-1 and tun-1-in-1 on node-3, respectively. Generated with a moderate debugging output. 

### Your configuration

```shell
config setup
    uniqueids=yes

conn %default
    keyingtries=%forever
    type=transport
    auto=route
    ike=aes_gcm256-sha2_256
    esp=aes_gcm256
    ikev2=insist
    retransmit-timeout=30

conn tun-2-in-1
    left=%defaultroute
    right=10.1.1.2
    leftid=@node-1
    rightid=@node-2
    leftcert="ovs_certkey_node-1"
    rightcert="ovs_cert_node-2"
    leftrsasigkey=%cert
    leftprotoport=udp/6081
    rightprotoport=udp

...

conn tun-3-in-1
    left=%defaultroute
    right=10.1.1.3
    leftid=@node-1
    rightid=@node-3
    leftcert="ovs_certkey_node-1"
    rightcert="ovs_cert_node-3"
    leftrsasigkey=%cert
    leftprotoport=udp/6081
    rightprotoport=udp

conn tun-3-out-1
    left=%defaultroute
    right=10.1.1.3
    leftid=@node-1
    rightid=@node-3
    leftcert="ovs_certkey_node-1"
    rightcert="ovs_cert_node-3"
    leftrsasigkey=%cert
    leftprotoport=udp
    rightprotoport=udp/6081
```

### Relevant log output

```shell
==== node-1 ====
Nov  8 10:27:04.463449: "tun-3-in-1": IKE SA proposals (connection add):
Nov  8 10:27:04.463466: "tun-3-in-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov  8 10:27:04.463478: "tun-3-in-1": Child SA proposals (connection add):
Nov  8 10:27:04.463482: "tun-3-in-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov  8 10:27:04.466135: "tun-3-in-1": loaded private key matching left certificate 'ovs_certkey_node-1'
Nov  8 10:27:04.467265: "tun-3-in-1": added IKEv2 connection
Nov  8 10:27:04.469080: "tun-3-in-1" #7: proposal 1:IKE=AES_GCM_C_256-HMAC_SHA2_256-MODP2048 chosen from remote proposals 1:IKE:ENCR=AES_GCM_C_256;PRF=HMAC_SHA2_256;DH=MODP2048;DH=MODP3072;DH=MODP4096;DH=MODP8192;DH=ECP_256;DH=ECP_384;DH=ECP_521;DH=CURVE25519[first-match]
Nov  8 10:27:04.472249: "tun-3-in-1" #8: initiating IKEv2 connection
Nov  8 10:27:04.474184: "tun-3-in-1" #7: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov  8 10:27:04.474257: "tun-3-in-1" #8: sent IKE_SA_INIT request to 10.1.1.3:500
Nov  8 10:27:04.482273: "tun-3-in-1" #8: sent IKE_AUTH request {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov  8 10:27:04.485573: "tun-3-in-1" #7: processing decrypted IKE_AUTH request: SK{IDi,CERT,CERTREQ,IDr,AUTH,SA,TSi,TSr,N(USE_TRANSPORT_MODE)}
Nov  8 10:27:04.485629: "tun-3-in-1" #7: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov  8 10:27:04.485864: "tun-3-in-1" #7: responder established IKE SA; authenticated peer '2048-bit RSASSA-PSS with SHA2_512' digital signature using preloaded certificate '@node-3' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-3'
Nov  8 10:27:04.488505: "tun-3-in-1" #10: No IKEv2 connection found with compatible Traffic Selectors
Nov  8 10:27:04.494539: "tun-3-out-1": IKE SA proposals (connection add):
Nov  8 10:27:04.494554: "tun-3-out-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov  8 10:27:04.494565: "tun-3-out-1": Child SA proposals (connection add):
Nov  8 10:27:04.494569: "tun-3-out-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov  8 10:27:04.495404: "tun-3-out-1": added IKEv2 connection
Nov  8 10:27:04.500068: "tun-3-out-1": queue Child SA; waiting on IKE SA "tun-3-in-1" #8 negotiating with 10.1.1.3
Nov  8 10:27:04.520149: "tun-3-in-1" #8: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov  8 10:27:04.520318: "tun-3-in-1" #8: initiator established IKE SA; authenticated peer '2048-bit RSASSA-PSS with SHA2_512' digital signature using preloaded certificate '@node-3' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-3'
Nov  8 10:27:04.550912: "tun-3-in-1" #9: initiator established Child SA using #8; IPsec transport [10.1.1.1-10.1.1.1:6081-6081 17] -> [10.1.1.3-10.1.1.3:0-65535 17] {ESP/ESN=>0x0bef6ee8 <0xbc60587b xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov  8 10:27:04.550961: "tun-3-out-1" #11: initiating Child SA using IKE SA #8
Nov  8 10:27:04.550970: "tun-3-out-1" #11: Child SA proposals (new child):
Nov  8 10:27:04.550975: "tun-3-out-1" #11:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov  8 10:27:04.551036: "tun-3-in-1" #12: Child SA proposals (new child):
Nov  8 10:27:04.551048: "tun-3-in-1" #12:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov  8 10:27:04.551076: "tun-3-in-1" #12: proposal 1:ESP=AES_GCM_C_256-MODP2048-ENABLED SPI=2ba1a736 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;DH=MODP2048;ESN=ENABLED;ESN
=DISABLED[first-match]
Nov  8 10:27:04.552769: "tun-3-out-1" #11: sent CREATE_CHILD_SA request for new IPsec SA
Nov  8 10:27:04.555309: "tun-3-in-1" #12: responder established Child SA using #8; IPsec transport [10.1.1.1-10.1.1.1:6081-6081 17] -> [10.1.1.3-10.1.1.3:0-65535 17] {ESP/ESN=>0x2ba1
a736 <0x5e503099 xfrm=AES_GCM_16_256-NONE-MODP2048 DPD=passive}
Nov  8 10:27:04.555390: "tun-3-out-1" #11: sent CREATE_CHILD_SA request for new IPsec SA
Nov  8 10:27:04.594837: EXPECTATION FAILED: ike->sa.st_v2_msgid_windows.responder.recv == ike->sa.st_v2_msgid_windows.responder.sent (is_duplicate_response() +1665 /programs/pluto/ik
ev2.c)
Nov  8 10:27:04.594849: "tun-3-in-1" #8: ISAKMP_v2_CREATE_CHILD_SA message response with Message ID 2 has no matching SA
Nov  8 10:27:05.055395: "tun-3-in-1" #8: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 0.5 seconds for response
Nov  8 10:27:05.055711: EXPECTATION FAILED: ike->sa.st_v2_msgid_windows.responder.recv == ike->sa.st_v2_msgid_windows.responder.sent (is_duplicate_response() +1665 /programs/pluto/ik
ev2.c)
Nov  8 10:27:05.055728: "tun-3-in-1" #8: ISAKMP_v2_CREATE_CHILD_SA message response with Message ID 2 has no matching SA
Nov  8 10:27:05.556315: "tun-3-in-1" #8: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 1 seconds for response
Nov  8 10:27:05.556570: EXPECTATION FAILED: ike->sa.st_v2_msgid_windows.responder.recv == ike->sa.st_v2_msgid_windows.responder.sent (is_duplicate_response() +1665 /programs/pluto/ik
ev2.c)
Nov  8 10:27:05.556579: "tun-3-in-1" #8: ISAKMP_v2_CREATE_CHILD_SA message response with Message ID 2 has no matching SA
Nov  8 10:27:06.557656: "tun-3-in-1" #8: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 2 seconds for response
Nov  8 10:27:06.557961: EXPECTATION FAILED: ike->sa.st_v2_msgid_windows.responder.recv == ike->sa.st_v2_msgid_windows.responder.sent (is_duplicate_response() +1665 /programs/pluto/ikev2.c)
Nov  8 10:27:06.557970: "tun-3-in-1" #8: ISAKMP_v2_CREATE_CHILD_SA message response with Message ID 2 has no matching SA

==== node-3 ====
Nov  8 10:27:04.460513: "tun-1-in-1": IKE SA proposals (connection add):
Nov  8 10:27:04.460529: "tun-1-in-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov  8 10:27:04.460541: "tun-1-in-1": Child SA proposals (connection add):
Nov  8 10:27:04.460545: "tun-1-in-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov  8 10:27:04.461960: "tun-1-in-1": added IKEv2 connection
Nov  8 10:27:04.466708: "tun-1-in-1" #2: initiating IKEv2 connection
Nov  8 10:27:04.468566: "tun-1-in-1" #2: sent IKE_SA_INIT request to 10.1.1.1:500
Nov  8 10:27:04.474357: "tun-1-in-1" #3: proposal 1:IKE=AES_GCM_C_256-HMAC_SHA2_256-MODP2048 chosen from remote proposals 1:IKE:ENCR=AES_GCM_C_256;PRF=HMAC_SHA2_256;DH=MODP2048;DH=MODP3072;DH=MODP4096;DH=MODP8192;DH=ECP_256;DH=ECP_384;DH=ECP_521;DH=CURVE25519[first-match]
Nov  8 10:27:04.476479: "tun-1-in-1" #3: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov  8 10:27:04.484001: "tun-1-in-1" #3: processing decrypted IKE_AUTH request: SK{IDi,CERT,CERTREQ,IDr,AUTH,SA,TSi,TSr,N(USE_TRANSPORT_MODE)}
Nov  8 10:27:04.484292: "tun-1-in-1" #3: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov  8 10:27:04.484515: "tun-1-in-1" #2: sent IKE_AUTH request {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov  8 10:27:04.484729: "tun-1-in-1" #3: responder established IKE SA; authenticated peer '2048-bit RSASSA-PSS with SHA2_512' digital signature using preloaded certificate '@node-1' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-1'
Nov  8 10:27:04.488248: "tun-1-out-1": IKE SA proposals (connection add):
Nov  8 10:27:04.488268: "tun-1-out-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov  8 10:27:04.488292: "tun-1-out-1": Child SA proposals (connection add):
Nov  8 10:27:04.488299: "tun-1-out-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov  8 10:27:04.489242: "tun-1-out-1": added IKEv2 connection
Nov  8 10:27:04.489433: "tun-1-in-1" #5: switched to "tun-1-out-1"
Nov  8 10:27:04.489448: "tun-1-out-1" #5: proposal 1:ESP=AES_GCM_C_256-ENABLED SPI=bc60587b chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;ESN=ENABLED;ESN=DISABLED[first-match]
Nov  8 10:27:04.519923: "tun-1-out-1" #5: responder established Child SA using #3; IPsec transport [10.1.1.3-10.1.1.3:0-65535 17] -> [10.1.1.1-10.1.1.1:6081-6081 17] {ESP/ESN=>0xbc60587b <0x0bef6ee8 xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov  8 10:27:04.520139: "tun-1-out-1" #6: initiating Child SA using IKE SA #3
Nov  8 10:27:04.520156: "tun-1-out-1" #6: Child SA proposals (new child):
Nov  8 10:27:04.520165: "tun-1-out-1" #6:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov  8 10:27:04.520251: "tun-1-in-1" #2: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov  8 10:27:04.520576: "tun-1-in-1" #2: initiator established IKE SA; authenticated peer '2048-bit RSASSA-PSS with SHA2_512' digital signature using preloaded certificate '@node-1' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-1'
Nov  8 10:27:04.520636: "tun-1-in-1" #4: IKE_AUTH response rejected Child SA with TS_UNACCEPTABLE
Nov  8 10:27:04.523854: "tun-1-out-1" #6: sent CREATE_CHILD_SA request for new IPsec SA
Nov  8 10:27:04.561786: "tun-1-in-1" #9: Child SA proposals (new child):
Nov  8 10:27:04.561795: "tun-1-in-1" #9:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov  8 10:27:04.561809: "tun-1-in-1" #9: proposal 1:ESP=AES_GCM_C_256-MODP2048-ENABLED SPI=b381b622 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;DH=MODP2048;ESN=ENABLED;ESN=DISABLED[first-match]
Nov  8 10:27:04.561872: "tun-1-in-1" #3: discarding packet received during asynchronous work (DNS or crypto) in STATE_V2_ESTABLISHED_IKE_SA
Nov  8 10:27:04.594721: EXPECTATION FAILED: ike->sa.st_connection->newest_ike_sa == ike->sa.st_serialno (process_v2_child_request_payloads() +395 /programs/pluto/ikev2_child.c)
Nov  8 10:27:04.594760: "tun-1-in-1" #9: responder established Child SA using #3; IPsec transport [10.1.1.3-10.1.1.3:6081-6081 17] -> [10.1.1.1-10.1.1.1:0-65535 17] {ESP/ESN=>0xb381b622 <0x1e4e597f xfrm=AES_GCM_16_256-NONE-MODP2048 DPD=passive}
Nov  8 10:27:05.055590: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:05.556503: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:06.557845: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:08.559697: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:12.562254: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:20.564881: "tun-1-in-1" #3: CREATE_CHILD_SA request has duplicate Message ID 2; retransmitting response
Nov  8 10:27:36.579317: "tun-1-in-1" #12: Child SA proposals (new child):
Nov  8 10:27:36.579355: "tun-1-in-1" #12:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov  8 10:27:36.579383: "tun-1-in-1" #12: switched to "tun-1-out-1"
Nov  8 10:27:36.579407: "tun-1-out-1" #12: proposal 1:ESP=AES_GCM_C_256-MODP2048-ENABLED SPI=e1690158 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;DH=MODP2048;ESN=ENABLED;ESN=DISABLED[first-match]
Nov  8 10:27:36.582219: "tun-1-out-1" #12: responder established Child SA using #2; IPsec transport [10.1.1.3-10.1.1.3:0-65535 17] -> [10.1.1.1-10.1.1.1:6081-6081 17] {ESP/ESN=>0xe1690158 <0x9e7b9fd1 xfrm=AES_GCM_16_256-NONE-MODP2048 DPD=passive}
Nov  8 10:27:40.548074: netlink_expire got message with length 60 < 232 bytes; ignore message
Nov  8 10:27:40.552481: netlink_expire got message with length 60 < 232 bytes; ignore message
Nov  8 10:27:40.552501: netlink_expire got message with length 60 < 232 bytes; ignore message
Nov  8 10:27:45.012896: Pluto is shutting down
Nov  8 10:27:45.013421: "tun-1-out-1" #12: ESP traffic information: in=0B out=0B
Nov  8 10:27:45.023392: ERROR: "tun-1-out-1" #6: netlink response for Del SA esp.2ba1a736@10.1.1.3: No such process (errno 3)
Nov  8 10:27:45.023446: "tun-1-out-1" #5: ESP traffic information: in=0B out=0B
Nov  8 10:27:45.039274: "tun-1-in-1" #2: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 40.572577s and sending notification
Nov  8 10:27:45.039429: "tun-1-in-1" #9: ESP traffic information: in=0B out=0B
Nov  8 10:27:45.047210: "tun-1-in-1" #3: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 40.572868s and sending notification
```

**Comment by cagney** [comment]
I'm not clear what exactly failed.

BTW, I suspect the PEXPECT:
```
EXPECTATION FAILED: ike->sa.st_connection->newest_ike_sa == ike->sa.st_serialno
```
is from 28e46f73859efe41af31a6d32544e2255fce5a20 which replaced:
```
ike->sa.st_connection->newest_ike_sa = ike->sa.st_serialno
```
in the create child sa code path; the assignment was bogus.  But it might paper over a crossing stream problem.

**Comment by cagney** [comment]
This is also of note 6341e0d0257f26a7883bc5d1abff50ac362c625b

**Comment by The-Mule** [comment]
So we bisected changes between 4.6 and 4.7 and actually realized that 1f4cede653c865d84ea4dd045db05670d513d385 is the first commit when we see the problem. If I get it right it is the initiator counterpart of the commit you mentioned. Do you recall any significant changes in this approach in v5?

Isn't it possible that both sides creates its own IKE SA because of a bad timing and hence they try to create their own Child SA as a part of it (when these commits enable that) and they end up without any working connections? Notice that what we do is running --start from all nodes to all other nodes with --asynchronous and hence it is very likely that some kind of crossing stream happens.

Also, after this commit when we hit the described problem,  we also start to see the following errors:

```
"tun-13-in-1" #29: ERROR: netlink response for Del SA esp.86e8a822@10.1.1.2 included errno 3: No such process
```

**Comment by The-Mule** [comment]
For instance, with 1f4cede653c865d84ea4dd045db05670d513d385 and a cluster of 20 nodes, there is no connection between nodes 5 and 12 and here's the log from 12 and 5, respectively:

```
# grep tun-12- tests/system-kmod-testsuite.dir/203/node-5/pluto.log
Nov 14 18:51:31.948402: "tun-12-in-1": IKE SA proposals:
Nov 14 18:51:31.948425: "tun-12-in-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov 14 18:51:31.948450: "tun-12-in-1": Child SA proposals:
Nov 14 18:51:31.948457: "tun-12-in-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov 14 18:51:31.950856: "tun-12-in-1": added IKEv2 connection
Nov 14 18:51:31.971951: "tun-12-in-1" #25: initiating IKEv2 connection
Nov 14 18:51:31.975079: "tun-12-in-1" #25: sent IKE_SA_INIT request
Nov 14 18:51:31.975915: "tun-12-in-1" #25: dropping unexpected IKE_SA_INIT message containing NO_PROPOSAL_CHOSEN notification; message payloads: N; missing payloads: SA,KE,Ni
Nov 14 18:51:32.221491: "tun-12-out-1": IKE SA proposals:
Nov 14 18:51:32.221515: "tun-12-out-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov 14 18:51:32.221535: "tun-12-out-1": Child SA proposals:
Nov 14 18:51:32.221542: "tun-12-out-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov 14 18:51:32.222927: "tun-12-out-1": added IKEv2 connection
Nov 14 18:51:32.230971: "tun-12-out-1": queue Child SA; waiting on IKE SA "tun-12-in-1" #25 negotiating with 10.1.1.12
Nov 14 18:51:32.476231: "tun-12-in-1" #25: STATE_V2_PARENT_I1: retransmission; will wait 0.5 seconds for response
Nov 14 18:51:32.476429: "tun-12-in-1" #25: dropping unexpected IKE_SA_INIT message containing NO_PROPOSAL_CHOSEN notification; message payloads: N; missing payloads: SA,KE,Ni
Nov 14 18:51:32.696888: "tun-12-out-1" #27: proposal 1:IKE=AES_GCM_C_256-HMAC_SHA2_256-MODP2048 chosen from remote proposals 1:IKE:ENCR=AES_GCM_C_256;PRF=HMAC_SHA2_256;DH=MODP2048;DH=MODP3072;DH=MODP4096;DH=MODP8192;DH=ECP_256;DH=ECP_384;DH=ECP_521;DH=CURVE25519[first-match]
Nov 14 18:51:32.700338: "tun-12-out-1" #27: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov 14 18:51:32.709021: "tun-12-out-1" #27: processing decrypted IKE_AUTH request: SK{IDi,CERT,CERTREQ,IDr,AUTH,SA,TSi,TSr,N}
Nov 14 18:51:32.709064: "tun-12-out-1" #27: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov 14 18:51:32.709430: "tun-12-out-1" #27: responder established IKE SA; authenticated using RSA with SHA2_512 and preloaded certificate '@node-12' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-12'
Nov 14 18:51:32.713926: "tun-12-out-1" #28: proposal 1:ESP=AES_GCM_C_256-ENABLED SPI=8c3f80b0 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;ESN=ENABLED;ESN=DISABLED[first-match]
Nov 14 18:51:32.829794: "tun-12-out-1" #28: responder established Child SA using #27; IPsec transport [10.1.1.5-10.1.1.5:0-65535 17] -> [10.1.1.12-10.1.1.12:6081-6081 17] {ESP/ESN=>0x8c3f80b0 <0xd3914d9e xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov 14 18:51:32.976097: "tun-12-out-1" #30: new Child SA:
Nov 14 18:51:32.976115: "tun-12-out-1" #30:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov 14 18:51:32.976131: "tun-12-out-1" #30: switched to "tun-12-in-1"
Nov 14 18:51:32.976148: "tun-12-in-1" #30: proposal 1:ESP=AES_GCM_C_256-MODP2048-ENABLED SPI=9322a690 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;DH=MODP2048;ESN=ENABLED;ESN=DISABLED[first-match]
Nov 14 18:51:32.979279: "tun-12-in-1" #25: STATE_V2_PARENT_I1: retransmission; will wait 1 seconds for response
Nov 14 18:51:33.101427: "tun-12-in-1" #30: responder established Child SA using #27; IPsec transport [10.1.1.5-10.1.1.5:6081-6081 17] -> [10.1.1.12-10.1.1.12:0-65535 17] {ESP/ESN=>0x9322a690 <0xfbe58664 xfrm=AES_GCM_16_256-NONE-MODP2048 DPD=passive}
Nov 14 18:51:33.101543: "tun-12-in-1" #25: state superseded by #30 try=1, drop this negotiation
Nov 14 18:51:33.101553: "tun-12-in-1" #25: encountered fatal error in state STATE_V2_PARENT_I1
Nov 14 18:51:33.101561: "tun-12-in-1" #25: deleting state (STATE_V2_PARENT_I1) aged 1.129621s and NOT sending notification
Nov 14 18:51:33.101674: "tun-12-in-1" #25: deleting IKE SA but connection is supposed to remain up; schedule EVENT_REVIVE_CONNS
Nov 14 18:51:33.101779: "tun-12-in-1": initiating connection 'tun-12-in-1' with serial $25 which received a Delete/Notify but must remain up per local policy
Nov 14 18:51:33.101809: "tun-12-in-1" #31: initiating Child SA using IKE SA #27
Nov 14 18:51:33.101818: "tun-12-in-1" #31: new Child SA:
Nov 14 18:51:33.101826: "tun-12-in-1" #31:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov 14 18:51:33.105163: "tun-12-in-1" #31: sent CREATE_CHILD_SA request for new IPsec SA
Nov 14 18:51:33.713982: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 0.5 seconds for response
Nov 14 18:51:34.401313: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 1 seconds for response
Nov 14 18:51:35.354859: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 2 seconds for response
Nov 14 18:51:37.356025: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 4 seconds for response
Nov 14 18:51:41.357976: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 8 seconds for response
Nov 14 18:51:49.364357: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: retransmission; will wait 16 seconds for response
Nov 14 18:52:05.369966: "tun-12-out-1" #27: STATE_V2_ESTABLISHED_IKE_SA: 30 second timeout exceeded after 6 retransmits.  No response (or no acceptable response) to our IKEv2 message
Nov 14 18:52:05.369987: "tun-12-out-1" #27: liveness action - putting connection into hold
Nov 14 18:52:05.370114: "tun-12-in-1" #31: ERROR: netlink response for Del SA esp.9a5f5331@10.1.1.5 included errno 3: No such process
Nov 14 18:52:05.370161: "tun-12-in-1" #30: ESP traffic information: in=0B out=0B
Nov 14 18:52:05.377828: "tun-12-out-1" #28: ESP traffic information: in=0B out=0B
Nov 14 18:52:05.385695: "tun-12-out-1" #27: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 32.688828s and NOT sending notification
Nov 14 18:52:05.385770: "tun-12-out-1" #27: deleting IKE SA but connection is supposed to remain up; schedule EVENT_REVIVE_CONNS
Nov 14 18:52:05.385821: "tun-12-out-1": initiating connection 'tun-12-out-1' with serial $26 which received a Delete/Notify but must remain up per local policy
Nov 14 18:52:05.385856: "tun-12-out-1" #98: initiating IKEv2 connection
Nov 14 18:52:05.387548: "tun-12-out-1" #98: sent IKE_SA_INIT request
Nov 14 18:52:05.394137: "tun-12-out-1" #98: sent IKE_AUTH request {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov 14 18:52:05.398462: "tun-12-out-1" #98: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov 14 18:52:05.398612: "tun-12-out-1" #98: initiator established IKE SA; authenticated using RSA with SHA2_512 and preloaded certificate '@node-12' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-12'
Nov 14 18:52:05.406812: "tun-12-out-1" #99: initiator established Child SA using #98; IPsec transport [10.1.1.5-10.1.1.5:0-65535 17] -> [10.1.1.12-10.1.1.12:6081-6081 17] {ESP/ESN=>0x51418516 <0xb50c6944 xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov 14 18:52:59.723330: "tun-12-out-1" #98: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 54.337479s and sending notification
Nov 14 18:52:59.723485: "tun-12-out-1" #99: ESP traffic information: in=0B out=456B
```

and 

```
# grep tun-5- tests/system-kmod-testsuite.dir/203/node-12/pluto.log
Nov 14 18:51:32.650802: "tun-5-in-1": IKE SA proposals:
Nov 14 18:51:32.650825: "tun-5-in-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov 14 18:51:32.650845: "tun-5-in-1": Child SA proposals:
Nov 14 18:51:32.650852: "tun-5-in-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov 14 18:51:32.653210: "tun-5-in-1": added IKEv2 connection
Nov 14 18:51:32.693553: "tun-5-in-1" #19: initiating IKEv2 connection
Nov 14 18:51:32.696778: "tun-5-in-1" #19: sent IKE_SA_INIT request
Nov 14 18:51:32.707114: "tun-5-in-1" #19: sent IKE_AUTH request {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov 14 18:51:32.851529: "tun-5-in-1" #19: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov 14 18:51:32.851791: "tun-5-in-1" #19: initiator established IKE SA; authenticated using RSA with SHA2_512 and preloaded certificate '@node-5' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-5'
Nov 14 18:51:32.956532: "tun-5-in-1" #20: initiator established Child SA using #19; IPsec transport [10.1.1.12-10.1.1.12:6081-6081 17] -> [10.1.1.5-10.1.1.5:0-65535 17] {ESP/ESN=>0xd3914d9e <0x8c3f80b0 xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov 14 18:51:32.956649: "tun-5-out-1": IKE SA proposals:
Nov 14 18:51:32.956662: "tun-5-out-1":   1:IKE=AES_GCM_C_256-HMAC_SHA2_256-NONE-MODP2048+MODP3072+MODP4096+MODP8192+ECP_256+ECP_384+ECP_521+CURVE25519
Nov 14 18:51:32.956681: "tun-5-out-1": Child SA proposals:
Nov 14 18:51:32.956688: "tun-5-out-1":   1:ESP=AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
Nov 14 18:51:32.958072: "tun-5-out-1": added IKEv2 connection
Nov 14 18:51:32.971939: "tun-5-out-1" #23: initiating Child SA using IKE SA #19
Nov 14 18:51:32.971960: "tun-5-out-1" #23: new Child SA:
Nov 14 18:51:32.971969: "tun-5-out-1" #23:   1:ESP=AES_GCM_C_256-NONE-MODP2048-ENABLED+DISABLED
Nov 14 18:51:32.976008: "tun-5-out-1" #23: sent CREATE_CHILD_SA request for new IPsec SA
Nov 14 18:51:32.979424: "tun-5-in-1" #24: proposal 1:IKE=AES_GCM_C_256-HMAC_SHA2_256-MODP2048 chosen from remote proposals 1:IKE:ENCR=AES_GCM_C_256;PRF=HMAC_SHA2_256;DH=MODP2048;DH=MODP3072;DH=MODP4096;DH=MODP8192;DH=ECP_256;DH=ECP_384;DH=ECP_521;DH=CURVE25519[first-match]
Nov 14 18:51:32.983112: "tun-5-in-1" #24: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov 14 18:51:33.133220: "tun-5-in-1" #19: discarding packet received during asynchronous work (DNS or crypto) in STATE_V2_ESTABLISHED_IKE_SA
Nov 14 18:51:33.259938: "tun-5-out-1" #23: initiator established Child SA using #19; IPsec transport [10.1.1.12-10.1.1.12:0-65535 17] -> [10.1.1.5-10.1.1.5:6081-6081 17] {ESP/ESN=>0xfbe58664 <0x9322a690 xfrm=AES_GCM_16_256-NONE-MODP2048 DPD=passive}
Nov 14 18:51:33.259982: "tun-5-in-1" #19: no STATE_V2_ESTABLISHED_IKE_SA message to send
Nov 14 18:51:33.714084: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:51:34.401631: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:51:35.354974: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:51:37.356169: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:51:41.358109: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:51:49.364526: "tun-5-in-1" #19: CREATE_CHILD_SA request has duplicate Message ID 0 but it is older than last response (2); message dropped
Nov 14 18:52:05.387886: "tun-5-in-1" #99: proposal 1:IKE=AES_GCM_C_256-HMAC_SHA2_256-MODP2048 chosen from remote proposals 1:IKE:ENCR=AES_GCM_C_256;PRF=HMAC_SHA2_256;DH=MODP2048;DH=MODP3072;DH=MODP4096;DH=MODP8192;DH=ECP_256;DH=ECP_384;DH=ECP_521;DH=CURVE25519[first-match]
Nov 14 18:52:05.390328: "tun-5-in-1" #99: sent IKE_SA_INIT reply {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_256 group=MODP2048}
Nov 14 18:52:05.395170: "tun-5-in-1" #99: processing decrypted IKE_AUTH request: SK{IDi,CERT,CERTREQ,IDr,AUTH,SA,TSi,TSr,N}
Nov 14 18:52:05.395224: "tun-5-in-1" #99: no Certificate Authority in NSS Certificate DB! certificate payloads discarded
Nov 14 18:52:05.395366: "tun-5-in-1" #99: switched to "tun-5-out-1"
Nov 14 18:52:05.395549: "tun-5-out-1" #99: responder established IKE SA; authenticated using RSA with SHA2_512 and preloaded certificate '@node-5' issued by CA 'C=US, ST=CA, L=Palo Alto, O=Open vSwitch, OU=Open vSwitch certifier, CN=node-5'
Nov 14 18:52:05.397856: "tun-5-out-1" #100: switched to "tun-5-in-1"
Nov 14 18:52:05.397883: "tun-5-in-1" #100: proposal 1:ESP=AES_GCM_C_256-ENABLED SPI=b50c6944 chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_256;ESN=ENABLED;ESN=DISABLED[first-match]
Nov 14 18:52:05.398284: "tun-5-in-1" #100: responder established Child SA using #99; IPsec transport [10.1.1.12-10.1.1.12:6081-6081 17] -> [10.1.1.5-10.1.1.5:0-65535 17] {ESP/ESN=>0xb50c6944 <0x51418516 xfrm=AES_GCM_16_256-NONE DPD=passive}
Nov 14 18:52:53.336451: "tun-5-out-1" #99: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 47.94858s and sending notification
Nov 14 18:52:53.336617: "tun-5-out-1" #23: ESP traffic information: in=0B out=456B
Nov 14 18:52:53.356795: "tun-5-in-1" #19: deleting state (STATE_V2_ESTABLISHED_IKE_SA) aged 80.663252s and sending notification
Nov 14 18:52:53.356920: "tun-5-in-1" #100: ESP traffic information: in=456B out=0B
Nov 14 18:52:53.364011: "tun-5-in-1" #24: deleting state (STATE_V2_PARENT_R1) aged 80.384605s and NOT sending notification
Nov 14 18:52:53.364118: "tun-5-in-1" #20: ESP traffic information: in=0B out=0B
Nov 14 18:52:53.364190: ERROR: "tun-5-in-1" #20: kernel: xfrm XFRM_MSG_DELPOLICYdelete(UNUSED) response for flow (in): No such file or directory (errno 2)
Nov 14 18:52:53.364198: "tun-5-in-1" #20: raw_policy in teardown_half_ipsec_sa() failed to delete inbound
```

There are a couple exchanges that not easy for me to understand (e.g. switching tun-X-in-1 to tun-X-out-1) but I suppose there will be more out of order that an experienced eye can catch.

**Comment by cagney** [comment]
> There are a couple exchanges that not easy for me to understand (e.g. switching tun-X-in-1 to tun-X-out-1) but I suppose there will be more out of order that an experienced eye can catch.

I'd assume that no one can fully explain some of the 4.x behaviour - I certainly couldn't.  It relied on the heuristic of picking the SA with the largest .so_serialno and it doesn't take much to construct scenarios where that fails (see the WIP crossing-stream-11- tests).

BTW, a 5.1 change to note is:
- keeping proper track of crossing Child SA exchanges
